### PR TITLE
parse_transform refactoring and other enhancements

### LINKED ISSUE #23: Recources across nodes
Resources are being cleaned up when the corresponding Erlang term is garbage collected. However, this does not work across nodes because of reference counting: the refcount reaches 0 for the slave node and it deallocates the resource. A mechanism is required that corrects the issue of resources being freed too early. It should also be able to free them as soon as they are not required anymore, not just when the whole context dies (for instance millions of statements being made and destroyed back to back) and perhaps it could also handle resources left behind in the wake of unexpected termination.

### LINKED ISSUE #27: dpi slave should be in hidden mode
slave node can be seen with `erlang:nodes()` and this is not ideal. This fun is used to do rpc multicall and should not hit the slave node

### PR DESCRIPTION
**Pull request by KarlKeiser**
fixes #23 
fixes #27 
Ref: https://github.com/K2InformaticsGmbH/oranif/issues/23#issuecomment-451900357

### CODE REVIEW

File: src/dpi_transform.erl [hunk 1]
```diff
@@ -68,8 +75,58 @@ parse_transform(Forms, Options) ->
...
+	  Replace =
+        fun(T) ->
+			case T of
+			  {function, _, fa, 2, _} -> New_fa;
+			  _Else -> T
+			end
+		end,    %% locates the fa and replaces it with the new one
+	  lists:map(Replace, Form); %% performs the replacement and returns the new document
+    _Else ->  %% no fa found, so make a new one
+	  Line = 0, % default line for new fa
```
Comment (c-bik) [inline src/dpi_transform.erl]: `fa` function MUST be added to the end of the code (so last line + 1)!

File: src/dpi_transform.erl [hunk 2]
```diff
@@ -68,8 +75,58 @@ parse_transform(Forms, Options) ->
...
+			      [{atom, Line + 1, fa_internal_unimplemented},
+			       {var, Line + 1, 'Fun'},
+			       {var, Line + 1, 'Args'}]}]}]}]}],
+
+    % try again, this time with a fresh fa template
+	insert_fa(Form ++ New_fa, NifDefns)
+end.
+
+%% takes a fa block as well as additional NifDefns and turns them into a new fa
+append_fa(Fa, NifDefns) ->
```
Comment (c-bik) [inline src/dpi_transform.erl]: Factor it into two functions
```erlang
NewFaClauses = make_fa_clauses(Line, NifDefs),
append_fa(Fa, NewFaClauses)
```

File: src/dpi_transform.erl [hunk 3]
```diff
@@ -68,8 +75,58 @@ parse_transform(Forms, Options) ->
...
+		   {clause, Line + 1,
+			[{atom, Line + 1, NameNif},
+		     cons(params(Args, Line), Line)],
+		    [],
+			[{call, Line + 2, {atom, Line + 2, NameNif},
+		      params(Args, Line)}]}
+		end,
+
+    %% build the list of new clauses
+    NewClauses = [ProcessTuple(X) || X <- NifDefns],
```
Comment (c-bik) [inline src/dpi_transform.erl]: Use `lists:foldl` instead local fun (` ProcessTuple = fun(Z) ...` etc)

File: src/dpi_transform.erl [hunk 4]
```diff
@@ -81,55 +138,123 @@ nif_exports(NifDefns, Line) ->
...
+nif_stubs([{Fun, FunNif, Arity, Args, Mode} | Rest],
+	  Line) ->
     FunParams = params(Args, Line + 1),
     FunGuards = guards(Args, FunParams, Line + 1),
     FunNifParams = params(Args, Line + 3),
     FunNifGuards = guards(Args, FunNifParams, Line + 3),
-    ?N("function ~p\n",[FunNif]),
+    ?N("function ~p\n", [FunNif]),
     [% -spec(Fun(Arg, ...) -> term()).
      {attribute, Line, spec,
```
Comment (c-bik) [inline src/dpi_transform.erl]: 1. Add a new line before every PT section
1. Comment the resulting erlang pseudocode from each pt block
1. Indent multiline terms (lists/tuple) with 4 spaces

File: src/dpi_transform.erl [hunk 5]
```diff
@@ -81,55 +138,123 @@ nif_exports(NifDefns, Line) ->
...
+	  [{call, Line + 2, {atom, Line + 2, get},
+	    [{atom, Line + 2, dpi_slave}]},
+	   {atom, Line + 2, dpi}, {atom, Line + 2, fa},
+	   {cons, Line + 2, {atom, Line + 2, FunNif},
+	    {cons, Line, cons(FunParams, Line + 1),
+	     {nil, Line}}}]}]}]},
+
+    %% generate auto version of the function. This is either the PRC or Gen
+    %% version depending on the mode of the function
+     {function, 96, Fun, Arity,
```
Comment (c-bik) [inline src/dpi_transform.erl]: Line numbers MUST always be calculated and **NEVER** hard coded!

File: src/dpi_transform.erl [hunk 6]
```diff
@@ -158,7 +283,7 @@ guards([{T, A} | Types], [P | Params], L) ->
      | guards(Types, Params, L)];
 guards([T | Types], [P | Params], L) ->
     [{call, L, {atom, L, guard(T)}, [P]} | guards(Types,
-                          Params, L)].
+						  Params, L)].
```
Comment (c-bik) [inline src/dpi_transform.erl]: Re indent!

File: src/dpi.erl [hunk 7]
```diff
@@ -6,7 +6,7 @@
 -export([load_unsafe/0]).
 
 % gen_server apis
--export([handle_call/3, handle_cast/2, init/1]).
+-export([handle_call/3, handle_cast/2, init/1, fa/2, stateCount/0]).
```
[the remark this answers is not in the document]
Comment (KarlKeiser) [reply src/dpi.erl]: `fa/2` needs to be exported because it is called using RPC.

**Pull request merged**